tools/resourcedocsgen/cmd/docs/registry.go

```diff
@@ -234,6 +307,115 @@ func resourceDocsFromRegistryCmd() *cobra.Command {
 	cmd.Flags().StringVar(&registryDir, "registryDir",
 		".",
 		"The root of the pulumi/registry directory")
+	cmd.Flags().BoolVar(&useAPI, "use-api", false, "Use the Pulumi Registry API instead of local files")
+	cmd.Flags().StringVar(&apiURL, "api-url",
+		"https://api.pulumi.com/api/preview/registry",
+		"URL of the Pulumi Registry API")
 
 	return cmd
 }
+
+// GetPackageMetadata implements PackageMetadataProvider for FileSystemProvider
+func (p *FileSystemProvider) GetPackageMetadata(pkgName string) (*pkg.PackageMeta, error) {
+	metadataFilePath := filepath.Join(getRegistryPackagesPath(p.registryDir), pkgName+".yaml")
+	b, err := os.ReadFile(metadataFilePath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading the metadata file %s", metadataFilePath)
+	}
+
+	var metadata pkg.PackageMeta
+	if err := yaml.Unmarshal(b, &metadata); err != nil {
+		return nil, errors.Wrapf(err, "unmarshalling the metadata file %s", metadataFilePath)
+	}
+
+	return &metadata, nil
+}
+
+// ListPackageMetadata implements PackageMetadataProvider for FileSystemProvider
+func (p *FileSystemProvider) ListPackageMetadata() ([]*pkg.PackageMeta, error) {
+	registryPackagesPath := getRegistryPackagesPath(p.registryDir)
+	files, err := os.ReadDir(registryPackagesPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading directory %s", registryPackagesPath)
+	}
+
+	// Count YAML files to pre-allocate the slice mostly to appease the linter.
+	var metadataCount int
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".yaml") {
+			metadataCount++
+		}
+	}
+
+	metadataList := make([]*pkg.PackageMeta, 0, metadataCount)
+	for _, file := range files {
+		if !strings.HasSuffix(file.Name(), ".yaml") {
+			continue
+		}
+
+		metadata, err := p.GetPackageMetadata(strings.TrimSuffix(file.Name(), ".yaml"))
+		if err != nil {
+			return nil, err
+		}
+		metadataList = append(metadataList, metadata)
+	}
+
+	return metadataList, nil
+}
+
+// GetPackageMetadata implements PackageMetadataProvider for RegistryAPIProvider
+func (p *RegistryAPIProvider) GetPackageMetadata(pkgName string) (*pkg.PackageMeta, error) {
+	resp, err := http.Get(fmt.Sprintf("%s/packages?name=%s", p.apiURL, pkgName))
+	if err != nil {
+		return nil, errors.Wrapf(err, "fetching package metadata from API for %s", pkgName)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("unexpected status code %d when fetching package metadata", resp.StatusCode)
+	}
+
+	var response PackageListResponse
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return nil, errors.Wrap(err, "decoding API response")
+	}
+
+	if len(response.Packages) == 0 {
+		return nil, errors.Errorf("no package found with name %s", pkgName)
+	}
+
+	if len(response.Packages) > 1 {
+		return nil, errors.Errorf("multiple packages found with name %s", pkgName)
+	}
+
+	return convertAPIPackageToPackageMeta(response.Packages[0])
```

Comment:
Nit: I'm a big fan of switch statements here, but it's 100% optional:
```suggestion
	switch len(response.Packages) {
	case 0:
		return nil, errors.Errorf("no package found with name %s", pkgName)
	case 1:
		return convertAPIPackageToPackageMeta(response.Packages[0])
	default:
		return nil, errors.Errorf("multiple packages found with name %s", pkgName)
	}
```

Reply:
i really like this pattern!